lambda/lambda.ml

```diff
@@ -674,79 +674,108 @@ let rec make_sequence fn = function
    Assumes that the image of the substitution is out of reach
    of the bound variables of the lambda-term (no capture). *)
 
-let subst update_env s lam =
-  let rec subst s lam =
-    let remove_list l s =
-      List.fold_left (fun s (id, _kind) -> Ident.Map.remove id s) s l
-    in
+let subst update_env ?(freshen_bound_variables = false) s lam =
+  (* [s] contains a partial substitution for the free variables of [lam];
+     during our traversal of the term we maintain a second environment
+     [l] with all the bound variables in scope, mapped to either themselves
+     or freshened versions of themselves when [freshen_bound_variables] is set.
+   *)
+  let bv id l =
+    let id' = if not freshen_bound_variables then id else Ident.rename id in
+    id', Ident.Map.add id id' l in
+  let bvs ids l =
+    List.fold_right (fun (id, kind) (ids', l) ->
+        let id', l = bv id l in
+        ((id', kind) :: ids' , l)
+      ) ids ([], l) in
+  let rec subst s l lam =
     match lam with
-    | Lvar id as l ->
-        begin try Ident.Map.find id s with Not_found -> l end
+    | Lvar id as lam ->
+        begin match Ident.Map.find_opt id l with
+          | Some id' -> Lvar id'
+          | None ->
+             (* note: as this point we know [id] is not a bound variable,
+                otherwise it would belong to [l]; it is a free variable. *)
+             begin match Ident.Map.find_opt id s with
+               | Some lam' -> lam'
+               | None -> lam
+             end
+        end
     | Lconst _ as l -> l
     | Lapply ap ->
-        Lapply{ap with ap_func = subst s ap.ap_func;
-                      ap_args = subst_list s ap.ap_args}
+        Lapply{ap with ap_func = subst s l ap.ap_func;
+                      ap_args = subst_list s l ap.ap_args}
     | Lfunction lf ->
-        let s =
-          List.fold_right
-            (fun (id, _) s -> Ident.Map.remove id s)
-            lf.params s
-        in
-        Lfunction {lf with body = subst s lf.body}
+        let params, l = bvs lf.params l in
+        Lfunction {lf with params; body = subst s l lf.body}
     | Llet(str, k, id, arg, body) ->
-        Llet(str, k, id, subst s arg, subst (Ident.Map.remove id s) body)
+        let id', l' = bv id l in
+        Llet(str, k, id', subst s l arg, subst s l' body)
```

Comment:
I started rewriting the whole function to follow the convention that `foo'` is the version of `foo` with substituted subexpressions, but this gets very verbose and very annoying to read fairly fast.

To read the definition above, I think of the following definition for a freshening function:

```
(let x = t in u)[sigma] = (let x' = t[sigma] in u[sigma, x'/x])   with x' fresh
```

`sigma` corresponds to our `l`, and `[sigma, x'/x]` to our `l'`.

Reply:
In the end I kept these cases as-is, but I changed `Lletrec` to use `l'` instead of `l`, so that `l` consistently refer to the outside environment, and `l'` to the inside environment (under the binder(s)).